src/Servers/Kestrel/Core/src/Internal/Infrastructure/Heartbeat.cs

```diff
@@ -71,17 +73,35 @@ internal void OnHeartbeat()
 
     private void TimerLoop()
     {
-        Thread.Sleep(_interval);
+        // Starting the heartbeat immediately triggers OnHeartbeat.
+        // Initial delay to avoid running heartbeat again from timer thread.
+        if (_stopWaitHandle.WaitOne(_interval))
+        {
+            return;
+        }
+
         while (!_stopped)
         {
             OnHeartbeat();
-            Thread.Sleep(_interval);
+
+            if (_stopWaitHandle.WaitOne(_interval))
+            {
+                return;
+            }
         }
     }
 
     public void Dispose()
     {
+        // Stop heart beat and immediately exit wait interval.
         _stopped = true;
-        // Don't block waiting for the thread to exit
+        _stopWaitHandle.Set();
+
+        // Wait for heartbeat thread to finish.
+        // Should either be immediate or a short delay while heartbeat callbacks complete.
+        if (_timerThread.IsAlive)
+        {
+            _timerThread.Join();
```

Comment:
The thread should exit quickly now that delay is interrupted. Removes the chance of heartbeat running while Kestrel is disposed.